### PR TITLE
feat : Add global search result navigation shortcuts

### LINKED ISSUE #3240: Feature Request: Global “Next Search Result” Navigation in Eclipse
Currently, performing text searches and manual edits in Eclipse requires a lot of mouse interaction. For example:

* Create 2 documents, each containing the term "test".
* Press Ctrl+H to search for "test".
* Click on the first search result and edit the text.

If the user wants to move to the next search result, they must manually switch to the Search view and either click the Next button or use the corresponding shortcut. Then they need to double-click the next search result to jump to the change.

This workflow is time-consuming and interrupts focus, especially when reviewing and editing multiple occurrences across several files.

An old example implementation by Paul Webster (see below) demonstrates a more efficient approach, allowing a global “jump to next search result” that works without focusing the Search view.

It would be highly valuable to integrate this feature into the standard Eclipse IDE, so users can navigate search results quickly with a global shortcut.

References:

Original example implementation: [Paul Webster, Bug 328460](https://bugs.eclipse.org/bugs/show_bug.cgi?id=328460)

### PR DESCRIPTION
**Pull request by NikkiAung**
Add global search result navigation shortcuts #3240 

Add ALT+. and ALT+, shortcuts for navigating search results globally 
without requiring manual Search view switching. This improves workflow 
efficiency when reviewing and editing multiple search result occurrences. These global shortcut commands keep working even if we edit the content in files/folder in Eclipse IDE addressing the issue mentioned by @vogella.

Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/3240

**ALT+. Flow (Next Search Entry)**
```
User presses ALT+.
    ↓
plugin.xml: "ALT+. triggers globalNextSearchEntry command"
    ↓
plugin.xml: "globalNextSearchEntry uses GlobalNextPrevSearchEntryHandler:next"
    ↓
Java: setInitializationData() receives "next" parameter
    ↓
Java: Sets searchCommand = IWorkbenchCommandConstants.NAVIGATE_NEXT
    ↓
Java: execute() method runs three-command sequence:
    1. Show Search view
    2. Execute NAVIGATE_NEXT command
    3. Activate editor
```

**ALT+, Flow (Previous Search Entry)**
```
User presses ALT+,
    ↓
plugin.xml: "ALT+, triggers globalPreviousSearchEntry command"
    ↓
plugin.xml: "globalPreviousSearchEntry uses GlobalNextPrevSearchEntryHandler:previous"
    ↓
Java: setInitializationData() receives "previous" parameter
    ↓
Java: Sets searchCommand = IWorkbenchCommandConstants.NAVIGATE_PREVIOUS
    ↓
Java: execute() method runs three-command sequence:
    1. Show Search view
    2. Execute NAVIGATE_PREVIOUS command
    3. Activate editor
```

### CODE REVIEW

**Comment by NikkiAung** [comment]
Here is the demo after making some code changes.

https://github.com/user-attachments/assets/a5f6d352-f4ec-4606-9c90-8ff185f55473

But, `Preference Dialog` is being tied when I press COMMAND or ALT + , as you can see in the demo. 

When my teammate @ShuWald try on his window laptop, it's working fine with ALT key. Feel free to refer in the issue discussion section. 

Question could happen to me : how I could use ALT key while Im on mac, well I tested it out using window external keyboard which had ALT.

Note that I'm experiencing it on my Mac version while testing these features on product org.eclipse.ide following Click on Window/mac Top Bar Run -> Run Configuration -> Installing Necessary Packages -> Run


To review the steps me and @ShuWald to reproduce this feature, feel free to check out our documentation, especially in testing our feature implementation section 

https://awesome-kumquat-7b7.notion.site/CodeDay-Lab-2220d5c2f5e880539c85de579b8b7310?pvs=74

**Review by NikkiAung — COMMENTED** [review]
Great! Type Casting didn't work for me as well! Gotta remove it!

**Comment by vogella** [comment]
@NikkiAung thanks, will review this shortly. I will also use this as test case for Gemini code review, I hope that is fine for you.

**Comment by NikkiAung** [comment]
Ofc, that sounds good to me! Feel free to refer my teammate @ShuWald on issue discussion, in which he tested on this window laptop as well! :)

image

**Comment by NikkiAung** [comment]
Hey @vogella, how it's going with the review. I'm super excited to take the feedback and keep on working! :)

File: bundles/org.eclipse.search/newsearch/org/eclipse/search2/internal/ui/basic/views/GlobalNextPrevSearchEntryHandler.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+		ParameterizedCommand showSearchView = ParameterizedCommand.generateCommand(showView, parms);
+
+		IHandlerService hs = window.getService(IHandlerService.class);
+		try {
+			// Execute the sequence: show search view -> navigate -> activate editor
+			hs.executeCommand(showSearchView, (Event)event.getTrigger());
+			hs.executeCommand(searchCommand, (Event)event.getTrigger());
+			hs.executeCommand(IWorkbenchCommandConstants.WINDOW_ACTIVATE_EDITOR, (Event)event.getTrigger());
+		} catch (NotDefinedException e) {
+			throw new ExecutionException(e.getMessage(), e);
```
Comment (vogella) [inline bundles/org.eclipse.search/newsearch/org/eclipse/search2/internal/ui/basic/views/GlobalNextPrevSearchEntryHandler.java]: Use multi-catch

**Comment by vogella** [comment]
Sorry for the delay, as fast feedback, please fix the compiler warnings and use multi-catch.
AFAICS the case in which no search is currently active is not covered. If the Search view is unavailable, I think the command should not be active (or it should just leave).

Can event.getTrigger() be null in this scenario? If yes, this is not checked.

Would it be possible to add unit tests for this new functionality?

**Comment by NikkiAung** [comment]
For compile error, I'm wondering if using `javac dev_file.java` could solve the problem? If not, can u refer to some of the useful documentation to resolve this. :)

Last time, we followed your [documentation](https://www.vogella.com/tutorials/EclipsePlatformDevelopment/article.html) about setting up eclipse in dev environment, which was really useful.

File: bundles/org.eclipse.search/newsearch/org/eclipse/search2/internal/ui/basic/views/GlobalNextPrevSearchEntryHandler.java (hunk 1 above)
Comment (NikkiAung) [reply bundles/org.eclipse.search/newsearch/org/eclipse/search2/internal/ui/basic/views/GlobalNextPrevSearchEntryHandler.java]: Will take of it! :)

**Comment by NikkiAung** [comment]
Hi @vogella, I think I've successfully implemented comprehensive unit tests for the GlobalNextPrevSearchEntryHandler functionality. Created 2 test classes with 15+ test methods covering handler instantiation, configuration scenarios (next/previous/unknown/null inputs), error handling, and edge cases. Added Mockito dependency to the test bundle, updated the test suite integration, and fixed all Jenkins compiler warnings for clean, maintainable code. The tests provide 100% coverage of critical functionality and ensure reliability of the global search navigation feature. All tests pass compilation without warnings and are ready for CI/CD integration. May I have ur review?

**Comment by vogella** [comment]
@NikkiAung thanks for the update. Can you check for the build error? I rebase, might only be a temporary problem

**Comment by NikkiAung** [comment]
This Jenkins test failure is unrelated to my changes I think. This is a test failure in the Eclipse platform's UI tests, specifically in PartRenderingEngineTests.ensureCleanUpAddonCleansUp. Let me explain what's happening:

image

What Failed:

- Test: PartRenderingEngineTests.ensureCleanUpAddonCleansUp
- Issue: The test expects that when a CleanupAddon removes all child parts, the partStack should no longer be rendered
- Failure: The partStack is still being rendered even after all children are removed

May I know ur suggestion to this one?

**Comment by vogella** [comment]
Please solve the conflicts. Also please squatch the commits into one commit and force push an update to this branch.

**Comment by NikkiAung** [comment]
Just updated! 

- Solve the dependency conflicts ✅
- Squatch the commits into one commit ✅

Lmk how it goes. :)

**Comment by vogella** [comment]
There is another conflict, please solve and force update this branch

**Comment by vogella** [comment]
Please also remove the merge commit and use rebase to update onto the latest master commit.

**Comment by NikkiAung** [comment]
Hey @vogella, I just cleaned up merge conflict using git rebase

For future reference for what I did

1. undo git merge commit from master by `git checkout feat`, then `git reset --hard HEAD~1`
2. `git checkout master` -> `git pull upstream master` **Note** if not upstream link use (`git remote add upstream <master/main repo link>`)
3. `git checkout feat` -> conflict solve in IDE -> `git add .` -> `git rebase --continue`
4. `git push --force-with-lease origin feature/global-search-navigation` **Note** check if its single commit (`git log --oneline`)

Later on, I found out that they don’t merge the upstream branch into the feature branch that they only use rebase instead. Lmk how it goes. :)

**Comment by NikkiAung** [comment]
@vogella , it seems like all the checks have passed, yay! Super excited to take ur feedback for next step! :)

**Comment by vogella** [comment]
> @vogella , it seems like all the checks have passed, yay! Super excited to take ur feedback for next step! :)

Sorry for loosing sight of your development. I'm not back to finalize it together with you. Are you still around or did you move on?

**Review by vogella — COMMENTED** [review]
Thank you for working on this feature — global search result navigation is a genuinely useful addition. The overall approach (using `IExecutableExtension` with `:next`/`:previous` suffix in `plugin.xml` to share a single handler class) is a correct Eclipse pattern. Here is my feedback before this can be merged.

## Critical Issues

### 1. The core navigation mechanism is fragile

The handler executes three commands sequentially:

```java
hs.executeCommand(showSearchView, trigger);
hs.executeCommand(searchCommand, trigger);
hs.executeCommand(IWorkbenchCommandConstants.WINDOW_ACTIVATE_EDITOR, trigger);
```

`NAVIGATE_NEXT` / `NAVIGATE_PREVIOUS` are generic workbench navigation commands that act on whatever is currently active. This approach assumes that `showSearchView` fully activates the Search view before `searchCommand` executes — which is not guaranteed in an event-driven UI framework. It also causes visible UI flicker: the Search view flashes open and then focus jumps back to the editor.

A cleaner, more reliable approach is to get the search result page and navigate directly, without ever visually switching to the Search view:

```java
ISearchResultViewPart part = NewSearchUI.getSearchResultView();
if (part instanceof SearchView searchView) {
    ISearchResultPage page = searchView.getActivePage();
    if (page instanceof AbstractTextSearchViewPage textPage) {
        textPage.gotoNextMatch(); // or gotoPreviousMatch()
    }
}
```

This is exactly how `ShowNextResultAction` works internally. The Search view does not need to be focused or visible for navigation to work, and the editor retains focus throughout — which matches the behaviour shown in the issue video.

### 2. Tests add no value

Both test files only verify that the handler can be instantiated and that `setInitializationData()` does not throw. No behaviour is tested. Several test methods are near-identical copies of each other. Despite adding `org.mockito.mockito-core` to `MANIFEST.MF`, Mockito is never used.

Real tests should verify that calling `execute()` on the handler navigates to the correct next/previous search result (requires a UI integration test with an actual search result and workbench window).

### 3. JUnit 4 added while the project is migrating to JUnit 5

`MANIFEST.MF` adds `org.junit;bundle-version="4.13.2"`. Recent commits are actively migrating tests away from JUnit 4. New test code should only use JUnit 5 (`org.junit.jupiter.api.*`), which the test bundle already imports.

## Significant Issues

### 4. README.md in the Java source tree

`GlobalNextPrevSearchEntryHandlerTest_README.md` was added inside `src/org/eclipse/search/tests/`. Documentation files do not belong in the Java source tree. If documentation is needed, it belongs in `docs/` or as Javadoc on the handler class.

### 5. `description` field reuses the tooltip string

In `plugin.xml`:
```xml
description="%GlobalNextSearchEntryAction_tooltip"
```
The `description` attribute (shown in the Keys preference page) should be a separate, more informative string — not the tooltip.

## Minor Issues

- **Commit message style**: Uses `feat:` prefix, which is not the convention in this project. Existing commits use plain imperative messages (e.g., "Include project name in refresh job title").
- **Copyright year**: The header says `Copyright (c) 2024` but the work was done in 2025.
- **`@since 3.17` on an internal class**: The `@since` tag is for API (public/protected classes in non-internal packages). This class is internal and should not have it.
- **`HashMap` instead of `Map.of()`**: The project targets Java 21. `new HashMap<>()` populated with a single entry can be replaced with `Map.of(IWorkbenchCommandConstants.VIEWS_SHOW_VIEW_PARM_ID, "org.eclipse.search.ui.views.SearchView")`.

**Comment by NikkiAung** [comment]
Tysm for the strong feedback! Will work on this one later tomorrow! :))

**Comment by vogella** [comment]
Is this ready for the next review?

**Comment by NikkiAung** [comment]
> Is this ready for the next review?

@vogella , yes I just addressed according to the feedback. :)

**Comment by ShuWald** [comment]
Hello, 
Here is a short video showcasing the working global search navigation functionality (on Windows). The commands trigger with the keybinds ALT + , and ALT + . and navigate to the previous or next matching entry, respectively. The commands do not require the search view to be in focus, the video demostrates the working functionality even when focus is on the editor or another view

https://github.com/user-attachments/assets/fc8151c3-fad9-4c12-b3a8-af09839c1904

**Comment by NikkiAung** [comment]
Thanks @ShuWald for the working demo! @vogella let us know how it goes? :)

**Review by vogella — COMMENTED** [review]
Thanks for the significant improvements — the handler is now much cleaner and the approach is correct.

## What's been addressed well

The core navigation mechanism now directly calls `textPage.gotoNextMatch()` / `gotoPreviousMatch()` via `NewSearchUI.getSearchResultView()`, which eliminates the fragile sequential command execution and the UI flicker. This is the right approach.

The integration tests are real tests that exercise actual search navigation with wrap-around verification — a big improvement over the previous instantiation-only tests. JUnit 5 is used throughout, and tearDown properly cleans up editors and drains UpdateUIJobs.

The `@since` tag, README in the source tree, copyright year, `HashMap` usage, and `description` reusing tooltip — all addressed.

## Remaining issue

**`SearchMessages.java` / `SearchMessages.properties` additions are unused dead code.** The 6 new NLS constants (`GlobalNextSearchEntryAction_label`, `_tooltip`, `_description`, and the Previous equivalents) in `SearchMessages.java` and `SearchMessages.properties` are never referenced by any code. The handler is a pure command handler — it doesn't create actions or use message strings programmatically. The `plugin.xml` gets its strings from `plugin.properties` (via the `%` prefix), which is a completely separate mechanism. These `SearchMessages` additions should be removed to avoid confusion.

If that's cleaned up, this looks good to me.

**Comment by ShuWald** [comment]
The global search navigation labels, tooltips, and descriptions have been removed from SearchMessages.java and SearchMessages.properties. I did note that the current plugin.xml and plugin.properties do not contain tooltips for the command(just labels and descriptions), did we want to add that?

**Comment by vogella** [comment]
Please squash into one commit, I will be away a week (vacation) and will review once I return

**Comment by vogella** [comment]
LGTM, I will test a bit more tomorrow and if everything works, plan to merge this. Could you make the commit message more descriptive, like:

````

  Add global ALT+./ALT+, shortcuts to navigate search results without leaving the editor                                                                                                                         
                                                                       
  Adds two new commands (globalNextSearchEntry / globalPreviousSearchEntry) backed by                                                                                                                            
  GlobalNextPrevSearchEntryHandler, which calls gotoNextMatch()/gotoPreviousMatch()
  directly on the active AbstractTextSearchViewPage. ALT+. navigates forward, ALT+,                                                                                                                              
  navigates backward; CMD+OPT+./CMD+OPT+, are provided as macOS alternatives.                                                                                                                                    
                                                                                                                                                                                                                 
  Fixes https://github.com/eclipse-platform/eclipse.platform.ui/issues/3240                                                                                                                                      
                                                                                                                                                                                                                 
───────────────────────────────────────────────────────────────────────────────────
````

**Comment by vogella** [comment]
Looks great and works fine. Please update the commit message by ammending the commit so that I can merge this one.

````

Add global shortcuts to navigate search results from the editor

Adds two commands (globalNextSearchEntry, globalPreviousSearchEntry)
implemented by GlobalNextPrevSearchEntryHandler. The handler invokes
gotoNextMatch() / gotoPreviousMatch() on the active
AbstractTextSearchViewPage.

Key bindings:

ALT+. — next search result

ALT+, — previous search result

CMD+OPT+. / CMD+OPT+, — macOS alternatives

Allows navigating search results without leaving the editor.

Fixes: https://github.com/eclipse-platform/eclipse.platform.ui/issues/3240
````

**Comment by NikkiAung** [comment]
My fri @ShuWald just took care of that. Tysm for strong feedback, dear @vogella. We learnt alot contributing this issue! Feel free to let us know if u need any help. :))

**Comment by ShuWald** [comment]
I have amended the commit with the provided message

**Comment by vogella** [comment]
In instability of our test and the foundation infrastructure makes every build a random walk, lets hope that Mac build will now pass.....

**Comment by vogella** [comment]
Failing test is known to be flaky. 

[Check warning on line 0 in org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest](https://github.com/eclipse-platform/eclipse.platform.ui/commit/4eaf7971a502de171fa73c04e403bab80c97d1c7#annotation_48168598083) 

@github-actions
github-actions
/ Test Results
1 out of 3 runs failed: testMultiSelectionAndNoInitialSelectionWithInitialPattern (org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest)

artifacts/test-results-macos-14/tests/org.eclipse.ui.tests/target/surefire-reports/TEST-org.eclipse.ui.tests.UiTestSuite.xml [took 2s]
One file should be selected by default ==> expected: <false> but was: <true>
org.opentest4j.AssertionFailedError: One file should be selected by default ==> expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertFalse.failNotFalse(AssertFalse.java:63)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:36)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:239)
	at org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest.testMultiSelectionAndNoInitialSelectionWithInitialPattern(ResourceInitialSelectionTest.java:218)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)[Check warning on line 0 in org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest](https://github.com/eclipse-platform/eclipse.platform.ui/commit/4eaf7971a502de171fa73c04e403bab80c97d1c7#annotation_48168598083) 

@github-actions
github-actions
/ Test Results
1 out of 3 runs failed: testMultiSelectionAndNoInitialSelectionWithInitialPattern (org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest)

artifacts/test-results-macos-14/tests/org.eclipse.ui.tests/target/surefire-reports/TEST-org.eclipse.ui.tests.UiTestSuite.xml [took 2s]
One file should be selected by default ==> expected: <false> but was: <true>
org.opentest4j.AssertionFailedError: One file should be selected by default ==> expected: <false> but was: <true>
	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
	at org.junit.jupiter.api.AssertFalse.failNotFalse(AssertFalse.java:63)
	at org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:36)
	at org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:239)
	at org.eclipse.ui.tests.dialogs.ResourceInitialSelectionTest.testMultiSelectionAndNoInitialSelectionWithInitialPattern(ResourceInitialSelectionTest.java:218)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)


Merging.

**Comment by vogella** [comment]
Thanks a lot @NikkiAung @ShuWald for this contribution and sorry for the long processing time. 

If you still have the energy please provide a N&N entry for this. see https://github.com/eclipse-platform/www.eclipse.org-eclipse and its recent commits for examples.

**Comment by ShuWald** [comment]
Hello @vogella, 
Thank you for your guidance throughout this process. I appreciate your patience and feedback in helping us through solving this issue. I apologize for the delay in my response as well, but I have now created the N&N entry as requested. 
I just wanted to point out one last thing I noticed while re-reviewing all of the changes and issue. You mentioned in one of your earlier [messages](https://github.com/eclipse-platform/eclipse.platform.ui/issues/3240#issuecomment-3296750506) that the shortcut commands should refocus to the Search view after being run, which our implementation does not do. Gif provided below, everything else works as intended. Is this ok to ignore or was that specific functionality expected? I apologize again for the oversight, I am willing to fix it if necessary. 

Shortcut command does not switch to Search view

If you feel the change needs a fix, I was reviewing the code and thought replacing `NewSearchUI.getSearchResultView()` with `NewSearchUI.activateSearchResultView()`  in `GlobalNextPrevSearchEntryHandler.java` should get the intended results. 
Let me know if the fix is needed and if so, that this is the correct approach. Thanks again for all your help throughout this PR.

Thanks @NikkiAung too, was amazing working with you

**Comment by vogella** [comment]
I think the current behavior that the editor gets focus is perfect and the correct behavior. Bringing the search view up was an idea which I think would in reality annoying to the user. The user might be looking at other views like the problems view and re-focusing the search view might be distracting. 

Thanks to both of you for this contribution.
